core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -582,22 +583,27 @@ public PlanOptimizers(
                                 new PushProjectionIntoTableScan(metadata, typeAnalyzer),
                                 new PushPredicateIntoTableScan(metadata, typeOperators, typeAnalyzer))));
 
+        ImmutableSet<Rule<?>> pushIntoTableScanRules = ImmutableSet.<Rule<?>>builder()
+                .addAll(columnPruningRules)
+                .addAll(projectionPushdownRules)
+                .add(new RemoveRedundantIdentityProjections())
+                .add(new PushLimitIntoTableScan(metadata))
+                .add(new PushPredicateIntoTableScan(metadata, typeOperators, typeAnalyzer))
+                .add(new PushSampleIntoTableScan(metadata))
+                .add(new PushJoinIntoTableScan(metadata))
+                .add(new PushAggregationIntoTableScan(metadata))
+                .add(new PushDistinctLimitIntoTableScan(metadata))
+                .add(new PushTopNIntoTableScan(metadata))
+                .build();
         IterativeOptimizer pushIntoTableScanOptimizer = new IterativeOptimizer(
                 ruleStats,
                 statsCalculator,
                 estimatedExchangesCostCalculator,
-                ImmutableSet.<Rule<?>>builder()
-                        .addAll(columnPruningRules)
-                        .addAll(projectionPushdownRules)
-                        .add(new RemoveRedundantIdentityProjections())
-                        .add(new PushLimitIntoTableScan(metadata))
-                        .add(new PushPredicateIntoTableScan(metadata, typeOperators, typeAnalyzer))
-                        .add(new PushSampleIntoTableScan(metadata))
-                        .add(new PushJoinIntoTableScan(metadata))
-                        .add(new PushAggregationIntoTableScan(metadata))
-                        .add(new PushDistinctLimitIntoTableScan(metadata))
-                        .add(new PushTopNIntoTableScan(metadata))
-                        .build());
+                pushIntoTableScanRules.stream()
+                        // PushJoinIntoTableScan must run only after ReorderJoins and DetermineJoinDistributionType to
+                        // ensure join related optimizations are applied before pushdown.
+                        .filter(rule -> !(rule instanceof PushJoinIntoTableScan))
```

Comment:
What if we added 
```
if (joinNode.getDistributionType().isEmpty()) {
    // ....
    return Result.empty();
}
```

to `PushJoinIntoTableScan` instead?

Reply:
See https://github.com/trinodb/trino/pull/7331#discussion_r596735504.

I could also change the `PATTERN` in PushJoinIntoTableScan to not match plans with an empty join distribution type.

Reply:
What about https://github.com/trinodb/trino/pull/7331#discussion_r596761482. I don't think join pushdown should depend on `DetermineJoinDistributionType` to be executed earlier.

Reply:
@sopel39 Fair point. `DetermineJoinDistributionType`'s output is no longer being used to avoid early join pushdown.
And in the case where join is pushed down into connectors the join node itself will be lost so it doesn't matter if `DetermineJoinDistributionType` ran or not for that node.

Reply:
> See #7331 (comment).

cc @losipiuk 

my point is -- the PlanOptimizers is a VERY complex class and I want to avoid making it much more complex, when possible.
I prefer to move complexity into rules instead.